### PR TITLE
Enforce code deprecation and virtual methods

### LINKED ISSUE #1620: Coding style changes: deprecated and override
I'm working on a PR to enforce use of `__ attribute__((deprecated))` throughout the framework; I'm proposing to use a macro `__deprecated` for convenience. The purpose is to ensure the framework (including samples) builds without internally referencing any deprecated items. Travis enforces this. It will also assist users in migrating their code.

The second item is to apply `override` to every overridden virtual function. The purpose is to catch typing errors; using virtual with a mis-spelled method name or parameter type will simply create a new method. Using `override` will throw a warning if the base method isn't declared `virtual`.

One possible point of contention is whether to use `virtual` or `override` on inherited destructors. I'm thinking not, as it doesn't add anything and can actually confuse. virtual destructors do not behave like regular virtual methods, and typos are already flagged by the compiler.

Any other 'modernisations' we can consider?

Thoughts?

### PR DESCRIPTION
**Pull request by mikee47**
* Use `SMING_DEPRECATED` attribute to ensure marked code not referenced within framework
* Apply `override` to all inherited virtual methods
* Ensure base virtual classes have virtual destructors
* Remove `virtual` from overriden destructors
* Remove empty overridden destructors and overriden methods that do not change base behaviour
* Ensure consistent ordering of constructors/destructor in header

Some samples use deprecated methods. If no suitable alternative has yet been provided, those methods have been un-deprecated and noted in the method comment using `@todo`. The follow methods have been un-deprecated completely:

	`HttpResponse::sendTemplate()`
	`WebsocketConnection::operator==(WebsocketConnection&)`

Additional style changes:

* Move trivial code into headers
* Use `nullptr` instead of `NULL` or `0` where appropriate
* Remove use of `__forceinline` or `inline` where there is no specific need
* Reduce sample applications `#include <SmingCore/*>` to `#include <*>`
* Fix ill-formed doxygen comments
* Fix doxygen build error and include SSL code in output

Closes #1620

### CODE REVIEW

**Comment by mikee47** [comment]
@slaff As ths PR deals with coding style changes I've gone further and updated all the core files to standardise comment headers and that #include guards are named correctly. As it affects unrelated files I've not committed it to this PR, but I can add it if you want.

**Comment by slaff** [comment]
> It affects unrelated files though so I'll assume it's a separate PR; I can commit here if you want.

As long as the `unrelated files` are not part of third-party libraries or Arduino libraries it should be fine to add it also here. But before you do this can you explain what you mean by 'unrelated files` ?

**Comment by mikee47** [comment]
`unrelated files` meaning files that otherwise wouldn't be modified as a result of deprecate/override

**Comment by slaff** [comment]
> `unrelated files` meaning files that otherwise wouldn't be modified as a result of deprecate/override

Separate PR please.

File: Sming/SmingCore/HardwarePWM.h [hunk 1]
```diff
@@ -64,8 +68,14 @@ class HardwarePWM
      *  @param  duty Value of duty cycle to set pin to
      *  @param  update Update PWM output
      *  @retval bool True on success
+	 *  @note   This function is used to set the pwm duty cycle for a given pin. If parameter 'update' is false
```
Comment (slaff) [inline Sming/SmingCore/HardwarePWM.h]: Please, indent the next two lines with the one above.

File: Sming/Wiring/Printable.h [hunk 2]
```diff
@@ -37,9 +37,11 @@
 #define PRINTABLE_H
 
 class Print;
+
 class Printable
 {
   public:
+	virtual ~Printable() { }
```
Comment (slaff) [inline Sming/Wiring/Printable.h]: The indentation needs fixing.

File: Sming/system/include/esp_systemapi.h [hunk 3]
```diff
@@ -32,6 +32,9 @@
  */
 #define SMING_UNUSED  __attribute__((unused))
 
+/* Flags a compiler warning when Sming framework methods, functions or types are changed */
+#define __deprecated __attribute__((deprecated))
```
Comment (slaff) [inline Sming/system/include/esp_systemapi.h]: The other definitions of attributes are uppercase. Can you make also this one to be __DEPRECATED or SMING_DEPRECATED ?

File: Sming/SmingCore/Network/Http/Websocket/WebsocketConnection.h [hunk 4]
```diff
@@ -121,9 +120,10 @@ class WebsocketConnection
 	void* getUserData();
 
 	// @deprecated
-	bool operator==(const WebsocketConnection& rhs) const;
+	bool operator==(const WebsocketConnection& rhs) const __deprecated;
```
Comment (mikee47) [inline Sming/SmingCore/Network/Http/Websocket/WebsocketConnection.h]: @slaff You deprecated this method on 04/05/2017, and enforcing deprecation means there's a corresponding change in the `HttpServer_WebSockets` sample and possibly user apps; Keep it deprecated?

File: Sming/system/include/esp_systemapi.h (hunk 3 above)
Comment (mikee47) [reply Sming/system/include/esp_systemapi.h]: I like SMING_DEPRECATED, it's specific

File: Sming/SmingCore/Network/Http/Websocket/WebsocketConnection.h (hunk 4 above)
Comment (slaff) [reply Sming/SmingCore/Network/Http/Websocket/WebsocketConnection.h]: > You deprecated this method on 04/05/2017, and enforc ....

Hm... cannot remember why. You can un-deprecated if needed.

**Comment by slaff** [comment]
@mikee47 Ready to merge?

File: Sming/SmingCore/Network/Http/HttpRequestAuth.h [hunk 5]
```diff
@@ -20,22 +20,25 @@ class HttpRequest;
...
+	virtual void setResponse(HttpResponse* response)
 	{
 	}
 };
 
 class HttpBasicAuth : public AuthAdapter
 {
 public:
-	HttpBasicAuth(const String& username, const String& password);
+	HttpBasicAuth(const String& username, const String& password)
```
Comment (slaff) [inline Sming/SmingCore/Network/Http/HttpRequestAuth.h]: @mikee47 You can initialize the variables in initialization list

image

File: Sming/SmingCore/Network/Http/HttpRequestAuth.h [hunk 6]
```diff
@@ -47,9 +50,16 @@ class HttpBasicAuth : public AuthAdapter
 class HttpDigestAuth : public AuthAdapter
 {
 public:
-	HttpDigestAuth(const String& username, const String& password);
+	HttpDigestAuth(const String& username, const String& password)
```
Comment (slaff) [inline Sming/SmingCore/Network/Http/HttpRequestAuth.h]: Here too: variables in initialization list

File: Sming/SmingCore/Network/Http/HttpRequest.h [hunk 7]
```diff
@@ -56,7 +56,8 @@ class HttpRequest
 		return new HttpRequest(*this);
 	}
 
-	HttpRequest& operator=(const HttpRequest& rhs);
+	/** @deprecated Please use `clone()` instead */
+	HttpRequest& operator=(const HttpRequest& rhs) SMING_DEPRECATED;
```
Comment (mikee47) [inline Sming/SmingCore/Network/Http/HttpRequest.h]: The copy operator provided is actually unsafe as it doesn't take copies of allocated member variables so I've marked it deprecated. I could go further and make it private, which would prevent applications building if it's used. Thoughts?

**Comment by mikee47** [comment]
Just noticed there's a bunch of overrides missing. Will add those.

File: Sming/SmingCore/Network/Http/HttpRequest.h (hunk 7 above)
Comment (slaff) [reply Sming/SmingCore/Network/Http/HttpRequest.h]: >  I could go further and make it private, which would prevent applications building if it's used. 

For now deprecate it only.

> I think probably get rid of the warning in the code and document the copy constructor instead.

Sounds good.

**Comment by mikee47** [comment]
@slaff Unless you find anything else that's all for this PR

**Comment by mikee47** [comment]
@slaff I'm inclined to revert the last change relating to HttpRequest copy constructor, deal with it separately. AFAIK it's only used by rBootHttpUpdate so functional changes may be necessary.

**Comment by slaff** [comment]
The build is failing:
```bash
SmingCore/Network/Http/HttpRequest.cpp: In copy constructor 'HttpRequest::HttpRequest(const HttpRequest&)':
SmingCore/Network/Http/HttpRequest.cpp:19:93: error: use of deleted function 'HttpParams::HttpParams(const HttpParams&)'
 ri), method(value.method), headers(value.headers), postParams(value.postParams)
                                                                               ^
In file included from SmingCore/Network/Http/HttpRequest.h:24:0,
                 from SmingCore/Network/Http/HttpRequest.cpp:13:
SmingCore/Network/Http/HttpParams.h:30:7: note: 'HttpParams::HttpParams(const HttpParams&)' is implicitly deleted because the default definition would be ill-formed:
 class HttpParams : public HashMap<String, String>, public Printable
       ^
In file included from SmingCore/Network/Http/HttpHeaders.h:23:0,
                 from SmingCore/Network/Http/HttpResponse.h:18,
                 from SmingCore/Network/Http/HttpRequestAuth.h:16,
                 from SmingCore/Network/Http/HttpRequest.h:18,
                 from SmingCore/Network/Http/HttpRequest.cpp:13:
Wiring/WHashMap.h:321:5: error: 'HashMap<K, V>::HashMap(const HashMap<K, V>&) [with K = String; V = String]' is private
     HashMap(const HashMap<K, V>& that);
     ^
In file included from SmingCore/Network/Http/HttpRequest.h:24:0,
                 from SmingCore/Network/Http/HttpRequest.cpp:13:
SmingCore/Network/Http/HttpParams.h:30:7: error: within this context
 class HttpParams : public HashMap<String, String>, public Printable
       ^
make[3]: *** [out/build/SmingCore/Network/Http//HttpRequest.o] Error 1
make[3]: Leaving directory `/home/travis/build/SmingHub/Sming/Sming'
```

> I'm inclined to revert the last change relating to HttpRequest copy constructor, deal with it separately

Yes, do it like this. This PR is already quite big.

**Comment by mikee47** [comment]
> This PR is already quite big.

Yes, and there shouldn't be any _functional_ changes in this PR.

**Comment by mikee47** [comment]
[Coding style rules](https://github.com/SmingHub/Sming/wiki/Coding-Style-Rules) updated

**Comment by slaff** [comment]
> [Coding style rules](https://github.com/SmingHub/Sming/wiki/Coding-Style-Rules) updated

Thanks for the changes. I have read the new text and all looks very good to me.

**Pull request merged**